File: wrinkle_deformer/wrinkle_deformer.py
import maya.OpenMaya as om
import maya.OpenMayaMPx as ommpx

class WrinkleDeformer(ommpx.MPxDeformerNode):
    kNodeName = "wrinkleDeformer"
    kNodeClassify = "deformer"
    kNodeID = om.MTypeId(0x100fff)

    aAmount = om.MObject()
    aDirection = om.MObject()

    def __init__(self):
        ommpx.MPxDeformerNode.__init__(self)

    def deform(self, data, itGeo, localToWorldMatrix, geomIndex):
        input = ommpx.cvar.MPxDeformerNode_input
        env = ommpx.cvar.MPxDeformerNode_envelope
        amount = data.inputValue(WrinkleDeformer.aAmount).asFloat()
        direction = data.inputValue(WrinkleDeformer.aDirection).asFloat()

        while not itGeo.isDone():
            index = itGeo.index()
            inputGeom = ommpx.MFnMesh(input)
            point = inputGeom.getPoint(index, om.MSpace.kObject)
            
            wrinkle = amount*(1-direction*point.y)
            point.x += wrinkle
            inputGeom.setPoint(index, point, om.MSpace.kObject)

            itGeo.next()
    
    def accessoryNodeSetup(self, dagMod):
        pass

    def accessoryNodeConnections(self):
        pass

def nodeCreator():
    return ommpx.asMPxPtr(WrinkleDeformer())

def nodeInitializer():
    nAttr = om.MFnNumericAttribute()

    WrinkleDeformer.aAmount = nAttr.create("amount", "amt", om.MFnNumericData.kFloat, 1.0)
    nAttr.setKeyable(True)
    WrinkleDeformer.addAttribute(WrinkleDeformer.aAmount)
    WrinkleDeformer.attributeAffects(WrinkleDeformer.aAmount, ommpx.cvar.MPxDeformerNode_outputGeom)

    WrinkleDeformer.aDirection = nAttr.create("direction", "dir", om.MFnNumericData.kFloat, 1.0)
    nAttr.setKeyable(True)
    WrinkleDeformer.addAttribute(WrinkleDeformer.aDirection)
    WrinkleDeformer.attributeAffects(WrinkleDeformer.aDirection, ommpx.cvar.MPxDeformerNode_outputGeom)

def initializePlugin(mobject):
    mplugin = om.MFnPlugin(mobject)
    try:
        mplugin.registerNode(WrinkleDeformer.kNodeName, WrinkleDeformer.kNodeID, nodeCreator, nodeInitializer, ommpx.MPxNode.kDeformerNode)
    except:
        sys.stderr.write("Failed to register node: %s" % WrinkleDeformer.kNodeName)
        raise

File: wrinkle_deformer/cpp/wrinkle_deformer.cpp
#include <maya/MFnNumericAttribute.h>
#include <maya/MFnMesh.h>
#include <maya/MItGeometry.h>
#include <maya/MPxDeformerNode.h>

class WrinkleDeformer : public MPxDeformerNode
{
public:
    WrinkleDeformer();
    virtual ~WrinkleDeformer();

    static void* creator();
    static MStatus initialize();

    virtual MStatus deform(MDataBlock& data, MItGeometry& itGeo, const MMatrix& localToWorldMatrix, unsigned int geomIndex);

    static MTypeId id;
    static MObject aAmount;
    static MObject aDirection;
};

MTypeId WrinkleDeformer::id(0x100000);
MObject WrinkleDeformer::aAmount;
MObject WrinkleDeformer::aDirection;

WrinkleDeformer::WrinkleDeformer() {}
WrinkleDeformer::~WrinkleDeformer() {}

void* WrinkleDeformer::creator()
{
    return new WrinkleDeformer();
}

MStatus WrinkleDeformer::initialize()
{
    MFnNumericAttribute nAttr;
    aAmount = nAttr.create("amount", "amt", MFnNumericData::kFloat, 1.0);
    nAttr.
#include <maya/MFnNumericAttribute.h>
#include <maya/MFnMesh.h>
#include <maya/MItGeometry.h>
#include <maya/MPxDeformerNode.h>

class WrinkleDeformer : public MPxDeformerNode
{
public:
    WrinkleDeformer();
    virtual ~WrinkleDeformer();

    static void* creator();
    static MStatus initialize();

    virtual MStatus deform(MDataBlock& data, MItGeometry& itGeo, const MMatrix& localToWorldMatrix, unsigned int geomIndex);

    static MTypeId id;
    static MObject aAmount;
    static MObject aDirection;
};

MTypeId WrinkleDeformer::id(0x100000);
MObject WrinkleDeformer::aAmount;
MObject WrinkleDeformer::aDirection;

WrinkleDeformer::WrinkleDeformer() {}
WrinkleDeformer::~WrinkleDeformer() {}

void* WrinkleDeformer::creator()
{
    return new WrinkleDeformer();
}

MStatus WrinkleDeformer::initialize()
{
    MFnNumericAttribute nAttr;
    aAmount = nAttr.create("amount", "amt", MFnNumericData::kFloat, 1.0);
    nAttr.setKeyable(true);
    addAttribute(aAmount);
    attributeAffects(aAmount, outputGeom);

    aDirection = nAttr.create("direction", "dir", MFnNumericData::kFloat, 1.0);
    nAttr.setKeyable(true);
    addattribute(aDirection);
    attributeAffects(aDirection, outputGeom);

    return MS::kSuccess;
}

MStatus WrinkleDeformer::deform(MDataBlock& data, MItGeometry& itGeo, const MMatrix& localToWorldMatrix, unsigned int geomIndex)
{
    MStatus status;

    float amount = data.inputValue(aAmount).asFloat();
    float direction = data.inputValue(aDirection).asFloat();

    MFnMesh fnMesh(inputGeom, &status);
    CHECK_MSTATUS_AND_RETURN_IT(status);

    for (; !itGeo.isDone(); itGeo.next()) {
        MPoint point = itGeo.position();
        float wrinkle = amount * (1 - direction * point.y);
        point.x += wrinkle;
        fnMesh.setPoint(itGeo.index(), point, MSpace::kObject);
    }

    return MS::kSuccess;
}
